hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/operation/PublisherCreateOperation.java

```diff
@@ -165,42 +162,77 @@ private QueryResult runInitialQuery() {
     }
 
     /**
-     * Replay the events on returned result-set which are generated during `runInitialQuery` call.
+     * Replay events over the result set of initial query. These events are
+     * received events during execution of the initial query.
      */
-    private void replayEventsOnResultSet(final QueryResult queryResult) throws Exception {
-        Collection<Object> resultCollection = readAccumulators();
-        for (Object result : resultCollection) {
-            if (result == null) {
+    private void replayEventsOverResultSet(QueryResult queryResult) throws Exception {
+        Map<Integer, Future<Object>> future = readAccumulators();
+        for (Map.Entry<Integer, Future<Object>> entry : future.entrySet()) {
+            Integer partitionId = entry.getKey();
+            Object eventsInOneAcc = entry.getValue().get();
+            if (eventsInOneAcc == null) {
                 continue;
             }
-            Object toObject = mapServiceContext.toObject(result);
-            List<QueryCacheEventData> eventDataList = (List<QueryCacheEventData>) toObject;
+            eventsInOneAcc = mapServiceContext.toObject(eventsInOneAcc);
+            List<QueryCacheEventData> eventDataList = (List<QueryCacheEventData>) eventsInOneAcc;
             for (QueryCacheEventData eventData : eventDataList) {
-                QueryResultRow entry = createQueryResultEntry(eventData);
-                add(queryResult, entry);
+                if (eventData.getDataKey() == null) {
+                    // this means a map-wide event like clear-all or evict-all
+                    // is inside the accumulator buffers
+                    removePartitionResults(queryResult, partitionId);
+                } else {
+                    add(queryResult, newQueryResultRow(eventData));
+                }
             }
         }
     }
 
-    private Collection<Object> readAccumulators() {
+    /**
+     * Remove matching entries from given result set with the given
+     * partition ID.
+     */
+    private void removePartitionResults(QueryResult queryResult, Integer partitionId) {
```

Comment:
Same here, an `Integer partitionId` looks strange. There are more candidates to check in this class.

Reply:
done